File: android/app/src/main/kotlin/fyi/yours/app/AppModel.kt
package fyi.yours.app

import android.app.Application
import android.os.Bundle
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import fyi.yours.app.net.ApiException
import fyi.yours.app.net.AuthFlow
import fyi.yours.app.net.SseEvent
import fyi.yours.app.net.TokenStore
import fyi.yours.app.net.YoursApi
import fyi.yours.app.support.MockData
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.UUID

// The client-side mirror of one resonance's state, and the verbs that move
// it. All persistence is server-side (see PROTOCOL.md); this object holds the
// bearer token, the loaded state, and the transient streaming UI state.
// Mirrors ios/Yours/AppModel.swift.
class AppModel(application: Application) : AndroidViewModel(application) {
    enum class Phase { LOADING, LANDING, CHAT, LOCKED, SLEEPING }

    data class DisplayMessage(
        val id: String = UUID.randomUUID().toString(),
        val role: String,
        val text: String,
        val isPulsing: Boolean = false,
        val isComplete: Boolean = true,
        val isError: Boolean = false
    )

    data class Notice(val message: String, val actionLabel: String, val action: NoticeAction)
    enum class NoticeAction { REFRESH, SIGN_IN }

    var phase by mutableStateOf(Phase.LOADING)
        private set
    var state by mutableStateOf<UniverseState?>(null)
        private set
    val messages = mutableStateListOf<DisplayMessage>()
    var composerText by mutableStateOf("")
    var isWaiting by mutableStateOf(false)
        private set
    var notice by mutableStateOf<Notice?>(null)
        private set
    var landingError by mutableStateOf<String?>(null)
        private set
    var sleepStartingTime by mutableStateOf<String?>(null)
        private set
    var settingsSubscription by mutableStateOf<SubscriptionDetails?>(null)
        private set

    private val prefs = application.getSharedPreferences("yours", Application.MODE_PRIVATE)
    var themePreference by mutableStateOf(ThemePreference.from(prefs.getString("yours-theme", null)))
        private set

    private val tokenStore = TokenStore(application)
    lateinit var api: YoursApi
        private set

    // Emulator screenshot / dev support, from MainActivity intent extras
    // (debug builds only): YoursMockChat, YoursMockLanding, YoursMockAutoSend,
    // YoursMockSleep, YoursToken, YoursBaseURL
    enum class MockMode { CHAT, LANDING }
    var mock: MockMode? = null
        private set
    private var autoSend = false
    private var mockSleep = false

    private var draftSaveJob: Job? = null
    private var pendingAuthVerifier: String?
        get() = prefs.getString("pending-auth-verifier", null)
        set(value) {
            prefs.edit().apply {
                if (value == null) remove("pending-auth-verifier")
                else putString("pending-auth-verifier", value)
            }.apply()
        }

    fun configure(extras: Bundle?) {
        if (this::api.isInitialized) return

        var baseUrlOverride: String? = null
        if (BuildConfig.DEBUG && extras != null) {
            if (extras.getBoolean("YoursMockChat")) mock = MockMode.CHAT
            if (extras.getBoolean("YoursMockLanding")) mock = MockMode.LANDING
            autoSend = extras.getBoolean("YoursMockAutoSend")
            mockSleep = extras.getBoolean("YoursMockSleep")
            extras.getString("YoursToken")?.let { tokenStore.token = it }
            baseUrlOverride = extras.getString("YoursBaseURL")
        }
        api = YoursApi(tokenStore.token, baseUrlOverride)

        viewModelScope.launch { start() }
    }

    private suspend fun start() {
        when (mock) {
            MockMode.CHAT -> {
                apply(MockData.state)
                if (autoSend) {
                    delay(1500)
                    composerText = "what does it look like from the other side?"
                    send()
                }
                if (mockSleep) beginSleep()
                return
            }
            MockMode.LANDING -> {
                phase = Phase.LANDING
                return
            }
            null -> {}
        }

        if (api.token == null) {
            phase = Phase.LANDING
            return
        }
        refreshState()
    }

    fun setTheme(preference: ThemePreference) {
        themePreference = preference
        prefs.edit().putString("yours-theme", preference.name).apply()
    }

    suspend fun refreshState() {
        if (mock != null) return
        try {
            apply(api.state())
        } catch (e: ApiException.Unauthenticated) {
            signOut("Your session has expired. Sign in to continue.")
        } catch (e: Exception) {
            if (phase == Phase.LOADING || phase == Phase.LANDING) {
                phase = Phase.LANDING
                landingError = "Couldn't reach Yours just now."
            } else {
                notice = Notice("Couldn't reach Yours just now.", "Try again", NoticeAction.REFRESH)
            }
        }
    }

    fun refresh() {
        viewModelScope.launch { refreshState() }
    }

    private fun apply(newState: UniverseState) {
        state = newState
        messages.clear()
        newState.narrative.forEach { messages.add(DisplayMessage(role = it.role, text = it.text)) }
        notice = null
        sleepStartingTime = null
        isWaiting = false
        landingError = null
        // Day 1 is free; day 2+ wants the subscription (policy is physics)
        phase = if (newState.universeDay > 1 && !newState.subscriptionActive) Phase.LOCKED else Phase.CHAT
        loadDraft()
    }

    // MARK: auth

    fun signIn() {
        val verifier = AuthFlow.newVerifier()
        pendingAuthVerifier = verifier
        AuthFlow.launch(getApplication(), api.baseUrl, AuthFlow.challenge(verifier))
    }

    // Called from MainActivity when the yours://auth?code=... intent arrives
    fun completeAuth(code: String) {
        val verifier = pendingAuthVerifier ?: return
        viewModelScope.launch {
            try {
                val (token, _) = api.exchangeToken(code, verifier)
                pendingAuthVerifier = null
                tokenStore.token = token
                api.token = token
                landingError = null
                refreshState()
            } catch (e: Exception) {
                landingError = "Sign-in didn't complete. Try again?"
                phase = Phase.LANDING
            }
        }
    }

    fun signOut(message: String? = null) {
        tokenStore.token = null
        api.token = null
        state = null
        messages.clear()
        composerText = ""
        phase = Phase.LANDING
        landingError = message
    }

    // MARK: chat

    fun send() {
        val text = composerText.trim()
        val universeTime = state?.universeTime ?: return
        if (text.isEmpty() || isWaiting) return

        notice = null
        messages.add(DisplayMessage(role = "user", text = text))
        composerText = ""
        isWaiting = true
        clearDraft()

        val index = messages.size
        messages.add(DisplayMessage(role = "assistant", text = ".", isPulsing = true, isComplete = false))

        viewModelScope.launch {
            if (mock != null) {
                MockData.streamResponse(this@AppModel, index)
                isWaiting = false
                return@launch
            }
            try {
                api.stream(ChatMessage.user(text), universeTime) { event -> handle(event, index) }
                messages[index] = messages[index].copy(isPulsing = false, isComplete = true)
            } catch (e: ApiException.Divergence) {
                messages.removeAt(index)
                notice = Notice(
                    e.serverMessage.ifEmpty { "This space moved forward elsewhere. Refresh to join where it is now." },
                    "Refresh to continue",
                    NoticeAction.REFRESH
                )
            } catch (e: ApiException.Unauthenticated) {
                messages.removeAt(index)
                notice = Notice(
                    "Your session has expired. Sign in to continue.",
                    "Sign in",
                    NoticeAction.SIGN_IN
                )
            } catch (e: ApiException.SubscriptionRequired) {
                messages.removeAt(index)
                refreshState()
            } catch (e: Exception) {
                messages[index] = messages[index].copy(
                    isPulsing = false,
                    isComplete = true,
                    isError = true,
                    text = "⚠️ Error: the stream broke. Your message wasn't lost on the web side — refresh to see where things stand."
                )
            }
            isWaiting = false
        }
    }

    private fun handle(event: SseEvent, index: Int) {
        when (event.name) {
            "message_start" -> replaceMessageText(index, "", pulsing = false)
            "content_block_delta" -> event.textDelta?.let { delta ->
                val current = if (messages[index].isPulsing) "" else messages[index].text
                replaceMessageText(index, current + delta, pulsing = false)
            }
            "message_stop" -> completeMessage(index)
            "universe_time" -> event.universeTime?.let { time ->
                state = state?.copy(universeTime = time)
            }
            "error" -> messages[index] = messages[index].copy(
                isPulsing = false,
                isComplete = true,
                isError = true,
                text = "⚠️ ${event.errorMessage ?: "An error occurred"}"
            )
            "end" -> messages[index] = messages[index].copy(isPulsing = false)
        }
    }

    fun replaceMessageText(index: Int, text: String, pulsing: Boolean) {
        messages[index] = messages[index].copy(text = text, isPulsing = pulsing)
    }

    fun completeMessage(index: Int) {
        messages[index] = messages[index].copy(isPulsing = false, isComplete = true)
    }

    fun noticeAction() {
        when (notice?.action) {
            NoticeAction.REFRESH -> refresh()
            NoticeAction.SIGN_IN -> signOut()
            null -> {}
        }
    }

    // MARK: draft (mirrors the web's localStorage + debounced server save)

    private val draftKey: String get() = "yours-input-${state?.universeTime ?: "current"}"

    private fun loadDraft() {
        val serverSaved = state?.textarea.orEmpty()
        val localSaved = prefs.getString(draftKey, null).orEmpty()
        // Whichever is longer is assumed more recent — same heuristic as the web
        val saved = if (serverSaved.length >= localSaved.length) serverSaved else localSaved
        if (saved.isNotEmpty()) composerText = saved
    }

    fun composerChanged() {
        prefs.edit().putString(draftKey, composerText).apply()

        draftSaveJob?.cancel()
        val text = composerText
        val key = draftKey
        val universeTime = state?.universeTime ?: return
        if (mock != null) return
        draftSaveJob = viewModelScope.launch {
            delay(1500)
            try {
                api.saveTextarea(text, universeTime)
                prefs.edit().remove(key).apply()
            } catch (e: ApiException.Divergence) {
                notice = Notice(
                    e.serverMessage.ifEmpty { "This space moved forward elsewhere. Refresh to join where it is now." },
                    "Refresh to continue",
                    NoticeAction.REFRESH
                )
            } catch (e: Exception) {
                // Fail silently — the draft is still in prefs
            }
        }
    }

    private fun clearDraft() {
        prefs.edit().remove(draftKey).apply()
        draftSaveJob?.cancel()
        val universeTime = state?.universeTime ?: return
        if (mock != null) return
        viewModelScope.launch {
            runCatching { api.saveTextarea("", universeTime) }
        }
    }

    // MARK: sleep

    suspend fun beginSleep() {
        if (mock != null) {
            sleepStartingTime = state?.universeTime
            phase = Phase.SLEEPING
            return
        }
        try {
            sleepStartingTime = api.beginSleep()
            phase = Phase.SLEEPING
        } catch (e: ApiException.SubscriptionRequired) {
            refreshState()
        } catch (e: Exception) {
            notice = Notice("Couldn't begin the night just now.", "Try again", NoticeAction.REFRESH)
        }
    }

    fun beginSleepAsync() {
        viewModelScope.launch { beginSleep() }
    }

    // Polled by SleepScreen once per second; true when integration completed
    suspend fun sleepIntegrationFinished(): Boolean {
        if (mock != null) return true
        val starting = sleepStartingTime ?: return true
        val current = runCatching { api.state() }.getOrNull() ?: return false
        return current.universeTime != starting
    }

    // MARK: settings

    fun loadSettingsSubscription() {
        if (mock != null) return
        viewModelScope.launch {
            settingsSubscription = runCatching { api.state(includeSubscription = true).subscription }.getOrNull()
        }
    }

    fun startOver() {
        if (mock != null) return
        viewModelScope.launch {
            try {
                api.reset()
                refreshState()
            } catch (e: Exception) {
                notice = Notice("Couldn't start over just now.", "Try again", NoticeAction.REFRESH)
            }
        }
    }

    suspend fun exportNarrative(): Pair<String, String> {
        val filename = "yours-${state?.universeTime?.replace(":", "-") ?: "current"}.txt"
        if (mock != null) return MockData.exportText to filename
        return api.exportText() to filename
    }
}

File: ios/Yours/AppModel.swift
import StoreKit
import SwiftUI

// The client-side mirror of one resonance's state, and the verbs that move
// it. All persistence is server-side (see PROTOCOL.md); this object holds the
// bearer token, the loaded state, and the transient streaming UI state.
@MainActor
final class AppModel: ObservableObject {
    enum Phase: Equatable {
        case loading
        case landing
        case chat
        case locked     // day 2+ without subscription — web sends you to settings
        case sleeping
    }

    struct DisplayMessage: Identifiable, Equatable {
        let id = UUID()
        var role: String
        var text: String
        var isPulsing = false
        var isComplete = true
        var isError = false
    }

    // The gentle full-width notices: continuity divergence, expired session
    struct Notice: Equatable {
        var message: String
        var actionLabel: String
        var action: NoticeAction
    }

    enum NoticeAction: Equatable {
        case refresh
        case signIn
    }

    @Published var phase: Phase = .loading
    @Published var state: UniverseState?
    @Published var messages: [DisplayMessage] = []
    @Published var composerText = ""
    @Published var isWaiting = false
    @Published var notice: Notice?
    @Published var landingError: String?
    @Published var sleepStartingTime: String?
    @Published var settingsSubscription: SubscriptionDetails?
    @Published var themePreference: ThemePreference {
        didSet { UserDefaults.standard.set(themePreference.rawValue, forKey: "yours-theme") }
    }

    let api: YoursAPI
    private let authFlow = AuthFlow()
    private var draftSaveTask: Task<Void, Never>?
    private var checkedExistingAppleEntitlement = false

    #if DEBUG
    // Simulator screenshot / preview support: -YoursMockChat, -YoursMockLanding
    var mock: MockMode?
    enum MockMode { case chat, landing }
    #endif

    let store = Store.shared

    init() {
        themePreference = ThemePreference(
            rawValue: UserDefaults.standard.string(forKey: "yours-theme") ?? "dark"
        ) ?? .dark
        #if DEBUG
        // Local-dev entry without the Google handshake: mint a token with
        // `bin/rails runner` and pass it as `-YoursToken <token>`
        if let injected = UserDefaults.standard.string(forKey: "YoursToken") {
            Keychain.token = injected
        }
        #endif
        api = YoursAPI(token: Keychain.token)
        #if DEBUG
        if ProcessInfo.processInfo.arguments.contains("-YoursMockChat") { mock = .chat }
        if ProcessInfo.processInfo.arguments.contains("-YoursMockLanding") { mock = .landing }
        #endif
        store.start()
    }

    // MARK: - Subscription (StoreKit)

    func subscribe(to product: Product) async {
        let accountToken = state?.iapAccountToken
        guard let newState = await store.purchase(product, accountToken: accountToken, api: api) else { return }
        apply(state: newState)
    }

    func restorePurchases() async {
        guard let state = await store.restore(api: api) else { return }
        apply(state: state)
    }

    // Permanent account deletion (App Store 5.1.1v). Destroys the resonance
    // server-side, then returns to the landing screen.
    func deleteAccount() async {
        #if DEBUG
        if mock != nil { signOut(); return }
        #endif
        do {
            try await api.deleteAccount()
            signOut()
        } catch {
            notice = Notice(message: "Couldn't delete the account just now.", actionLabel: "Try again", action: .refresh)
        }
    }

    // MARK: - Lifecycle

    func start() async {
        #if DEBUG
        switch mock {
        case .chat:
            apply(state: MockData.state)
            if ProcessInfo.processInfo.arguments.contains("-YoursMockAutoSend") {
                try? await Task.sleep(for: .seconds(1.5))
                composerText = "what does it look like from the other side?"
                send()
            }
            if ProcessInfo.processInfo.arguments.contains("-YoursMockSleep") {
                await beginSleep()
            }
            return
        case .landing:
            phase = .landing
            return
        case nil:
            break
        }
        #endif

        guard api.token != nil else {
            phase = .landing
            return
        }
        await refreshState()
    }

    func refreshState() async {
        #if DEBUG
        if mock != nil { return }
        #endif
        do {
            let state = try await api.state()
            let syncedState = await stateAfterSyncingExistingEntitlementIfNeeded(state)
            apply(state: syncedState)
        } catch APIError.unauthenticated {
            signOut(message: "Your session has expired. Sign in to continue.")
        } catch {
            if phase == .loading || phase == .landing {
                phase = .landing
                landingError = "Couldn't reach Yours just now."
            } else {
                notice = Notice(
                    message: "Couldn't reach Yours just now.",
                    actionLabel: "Try again",
                    action: .refresh
                )
            }
        }
    }

    private func stateAfterSyncingExistingEntitlementIfNeeded(_ state: UniverseState) async -> UniverseState {
        guard state.universeDay > 1,
              !state.subscriptionActive,
              !checkedExistingAppleEntitlement
        else { return state }

        checkedExistingAppleEntitlement = true
        if let restored = await store.syncExistingEntitlement(api: api) {
            return restored
        }
        return state
    }

    private func apply(state: UniverseState) {
        self.state = state
        messages = state.narrative.map { DisplayMessage(role: $0.role, text: $0.text) }
        notice = nil
        sleepStartingTime = nil
        isWaiting = false
        landingError = nil
        // Day 1 is free; day 2+ wants the subscription (policy is physics)
        phase = (state.universeDay > 1 && !state.subscriptionActive) ? .locked : .chat
        loadDraft()
    }

    // MARK: - Auth

    func signIn() async {
        do {
            let result = try await authFlow.signIn(api: api)
            Keychain.token = result.token
            api.token = result.token
            landingError = nil
            await refreshState()
        } catch is AuthFlow.Cancelled {
            // The human closed the sheet; nothing to say
        } catch {
            landingError = "Sign-in didn't complete. Try again?"
        }
    }

    func handleOpenURL(_ url: URL) {
        _ = authFlow.handleCallbackURL(url)
    }

    func signOut(message: String? = nil) {
        Keychain.token = nil
        api.token = nil
        checkedExistingAppleEntitlement = false
        state = nil
        messages = []
        composerText = ""
        phase = .landing
        landingError = message
    }

    // MARK: - Chat

    func send() {
        let text = composerText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, !isWaiting, let universeTime = state?.universeTime else { return }

        notice = nil
        messages.append(DisplayMessage(role: "user", text: text))
        composerText = ""
        isWaiting = true
        clearDraft()

        // Track the streaming message by its stable id, never by index: a
        // concurrent apply(state:) (from a refresh, a purchase, a mid-stream
        // universe_time) can replace the whole array, and the error paths below
        // remove messages — any captured index would go stale and misroute or
        // trap. updateStreaming(_:) looks the message up by id each time and
        // no-ops if it's gone.
        let placeholder = DisplayMessage(role: "assistant", text: ".", isPulsing: true, isComplete: false)
        let streamingID = placeholder.id
        messages.append(placeholder)

        Task {
            #if DEBUG
            if mock != nil {
                await MockData.streamResponse(into: self, id: streamingID)
                isWaiting = false
                return
            }
            #endif
            do {
                let events = try await api.stream(message: .user(text), universeTime: universeTime)
                for try await event in events {
                    handle(event, id: streamingID)
                }
                updateStreaming(streamingID) { $0.isPulsing = false; $0.isComplete = true }
            } catch APIError.divergence(let message) {
                removeMessage(streamingID)
                notice = Notice(
                    message: message.isEmpty
                        ? "This space moved forward elsewhere. Refresh to join where it is now."
                        : message,
                    actionLabel: "Refresh to continue",
                    action: .refresh
                )
            } catch APIError.unauthenticated {
                removeMessage(streamingID)
                notice = Notice(
                    message: "Your session has expired. Sign in to continue.",
                    actionLabel: "Sign in",
                    action: .signIn
                )
            } catch APIError.subscriptionRequired {
                removeMessage(streamingID)
                await refreshState()
            } catch {
                updateStreaming(streamingID) {
                    $0.isPulsing = false
                    $0.isComplete = true
                    $0.isError = true
                    $0.text = "⚠️ Error: the stream broke. Your message wasn't lost on the web side — refresh to see where things stand."
                }
            }
            isWaiting = false
        }
    }

    // Mutate the in-flight streaming message by id; no-op if it's no longer in
    // the array (e.g. an apply(state:) replaced it). This is the guard that
    // makes streaming safe against concurrent state replacement.
    func updateStreaming(_ id: UUID, _ mutate: (inout DisplayMessage) -> Void) {
        guard let i = messages.firstIndex(where: { $0.id == id }) else { return }
        mutate(&messages[i])
    }

    private func removeMessage(_ id: UUID) {
        messages.removeAll { $0.id == id }
    }

    private func handle(_ event: SSEEvent, id: UUID) {
        switch event.name {
        case "message_start":
            updateStreaming(id) { $0.isPulsing = false; $0.text = "" }
        case "content_block_delta":
            if let delta = event.textDelta {
                updateStreaming(id) {
                    if $0.isPulsing { $0.isPulsing = false; $0.text = "" }
                    $0.text += delta
                }
            }
        case "message_stop":
            updateStreaming(id) { $0.isComplete = true }
        case "universe_time":
            if let time = event.universeTime {
                state?.universeTime = time
            }
        case "error":
            updateStreaming(id) {
                $0.isPulsing = false
                $0.isComplete = true
                $0.isError = true
                $0.text = "⚠️ \(event.errorMessage ?? "An error occurred")"
            }
        case "end":
            updateStreaming(id) { $0.isPulsing = false }
        default:
            break
        }
    }

    func noticeAction() {
        guard let notice else { return }
        switch notice.action {
        case .refresh:
            Task { await refreshState() }
        case .signIn:
            signOut()
        }
    }

    // MARK: - Draft (mirrors the web's localStorage + debounced server save)

    private var draftKey: String {
        "yours-input-\(state?.universeTime ?? "current")"
    }

    private func loadDraft() {
        let serverSaved = state?.textarea ?? ""
        let localSaved = UserDefaults.standard.string(forKey: draftKey) ?? ""
        // Whichever is longer is assumed more recent — same heuristic as the web
        let saved = serverSaved.count >= localSaved.count ? serverSaved : localSaved
        if !saved.isEmpty {
            composerText = saved
        }
    }

    func composerChanged() {
        UserDefaults.standard.set(composerText, forKey: draftKey)

        draftSaveTask?.cancel()
        let text = composerText
        let key = draftKey
        guard let universeTime = state?.universeTime else { return }
        draftSaveTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(1.5))
            guard !Task.isCancelled, let self else { return }
            #if DEBUG
            if self.mock != nil { return }
            #endif
            do {
                try await self.api.saveTextarea(text, universeTime: universeTime)
                UserDefaults.standard.removeObject(forKey: key)
            } catch APIError.divergence(let message) {
                self.notice = Notice(
                    message: message.isEmpty
                        ? "This space moved forward elsewhere. Refresh to join where it is now."
                        : message,
                    actionLabel: "Refresh to continue",
                    action: .refresh
                )
            } catch {
                // Fail silently — the draft is still in UserDefaults
            }
        }
    }

    private func clearDraft() {
        UserDefaults.standard.removeObject(forKey: draftKey)
        draftSaveTask?.cancel()
        guard let universeTime = state?.universeTime else { return }
        #if DEBUG
        if mock != nil { return }
        #endif
        Task { try? await api.saveTextarea("", universeTime: universeTime) }
    }

    // MARK: - Sleep

    func beginSleep() async {
        #if DEBUG
        if mock != nil {
            sleepStartingTime = state?.universeTime
            phase = .sleeping
            return
        }
        #endif
        do {
            sleepStartingTime = try await api.beginSleep()
            phase = .sleeping
        } catch APIError.subscriptionRequired {
            await refreshState()
        } catch {
            notice = Notice(
                message: "Couldn't begin the night just now.",
                actionLabel: "Try again",
                action: .refresh
            )
        }
    }

    // Polled by SleepView once per second; true when integration completed
    func sleepIntegrationFinished() async -> Bool {
        #if DEBUG
        if mock != nil { return true }
        #endif
        guard let starting = sleepStartingTime else { return true }
        guard let current = try? await api.state() else { return false }
        return current.universeTime != starting
    }

    // MARK: - Settings

    func loadSettingsSubscription() async {
        #if DEBUG
        if mock != nil { return }
        #endif
        settingsSubscription = try? await api.state(includeSubscription: true).subscription
    }

    func startOver() async {
        #if DEBUG
        if mock != nil { return }
        #endif
        do {
            try await api.reset()
            await refreshState()
        } catch {
            notice = Notice(message: "Couldn't start over just now.", actionLabel: "Try again", action: .refresh)
        }
    }

    func exportNarrative() async throws -> (text: String, filename: String) {
        let filename = "yours-\(state?.universeTime.replacingOccurrences(of: ":", with: "-") ?? "current").txt"
        #if DEBUG
        if mock != nil { return (MockData.exportText, filename) }
        #endif
        return (try await api.exportText(), filename)
    }
}
